website/blog/2021-11-11-running-gvisor-in-production-at-scale-in-ant.md

```diff
@@ -0,0 +1,412 @@
+<code>futex()</code> is introduced by timers (64 timer buckets), and that Sentry
+syscalls walks a much longer path (redpill), makes it worse.
+
+We have two optimizations here: 1. decrease the number of timer buckets, from 64
+to 4; 2. decrease the timer precision from ns to ms. You may worry about the
+decrease of timer precision, but as we see, most of the applications are
+event-based, and not affected by a coarse grained timer.
+
+However, Go changes the implementation of timer in v1.14; how to port this
+optimization remains an open question.
+
+### Scheduler
+
+gVisor introduces an extra level of schedule along with the host linux
+scheduler (usually CFS). A L2 scheduler sometimes brings positive impact as it
+saves the heavy context switch in the L1 scheduler. We can find many two-level
+scheduler cases, for example, coroutines, virtual machines, etc.
+
+gVisor reuses Go's work-stealing scheduler, which is originally designed for
+coroutines, as the L2 scheduler. They share the same goal:
+
+"We need to balance between keeping enough running worker threads to utilize
+available hardware parallelism and parking excessive running worker threads
+to conserve CPU resources and power." -- From
+[Go scheduler code](https://golang.org/src/runtime/proc.go).
+
+If not properly tuned, the L2 scheduler may leak the schedule pressure to the
+L1 scheduler. According to G-P-M model of Go, the parallelism is close related to
+the GOMAXPROCS limit. The upstream gVisor by default uses # of host cores,
+which leads to a lot of wasted M wake/stop(s). By properly configuring the
+GOMAXPROCS of a POD of 4/8/16 cores, we find it can save some CPU cycles
+without worsening the workload latency.
+
+To further restrict extra M wake/stop(s), before wakep(), we calculate the # of
+running Gs and # of running Ps to decide if necessary to wake a M. And we find
+it's better to firstly steal from the longest local run queue, comparing to
+previously random-sequential way. Another related optimization is that we find
+most applications will get back to Sentry very soon, and it's not necessary to
+handle off its P when it leaves into user space and find an idle P when it gets
+back.
+
+Some optimizations in Go are put [here](https://github.com/zhuangel/go/tree/go1.13.4.blog).
+What we learned from the optimization process of gVisor is that digging into
+Go runtime to understand what's going on there. And it's normal that some ideas
+work, but some fail.
+
+## Summary
+
+We introduced how we profiled gVisor for production-ready performance. Using
+this methodology, along with some other aggressive measures, we finally got to
+run gVisor with an acceptable overhead, and even better than runc in some
+workloads.  We also absorbed a lot of optimization progress in the community,
+e.g., VFS2.
+
+So far, we have deployed more than 100K gVisor instances in the production
+environment. And it very well supported transactions of 11.11 Global Shopping
```

Comment:
s/11.11/Singles Day/

With link to https://en.wikipedia.org/wiki/Singles%27_Day